plutus-scb/src/Cardano/Protocol.hs

```diff
@@ -0,0 +1,212 @@
+{-# LANGUAGE AllowAmbiguousTypes #-}
+{-# LANGUAGE DataKinds           #-}
+{-# LANGUAGE DeriveAnyClass      #-}
+{-# LANGUAGE DerivingVia         #-}
+{-# LANGUAGE FlexibleContexts    #-}
+{-# LANGUAGE FlexibleInstances   #-}
+{-# LANGUAGE LambdaCase          #-}
+{-# LANGUAGE OverloadedStrings   #-}
+{-# LANGUAGE RankNTypes          #-}
+{-# LANGUAGE TypeFamilies        #-}
+{-# LANGUAGE TypeOperators       #-}
+{-# OPTIONS -fno-warn-orphans #-}
+
+module Cardano.Protocol where
+
+import           Codec.Serialise.Class           (Serialise)
+import           Data.Aeson                      (FromJSON, ToJSON)
+import qualified Data.ByteString.Lazy            as BSL
+import           Data.Functor                    (void)
+import           Data.List                       (findIndex)
+import           Data.Text.Prettyprint.Doc       (Pretty)
+import           Eventful.Projection             (GlobalStreamProjection, Projection (..), StreamProjection (..))
+import           Eventful.Store.Class
+import           GHC.Generics
+
+import           Control.Lens                    hiding (index, modifying, use)
+import           Control.Monad.Freer             (Eff, Member, Members, type (~>), interpret, send)
+import           Control.Monad.Freer.Extra.Log
+import           Control.Monad.Freer.Extra.State
+import           Control.Monad.Freer.State       as Eff
+import           Control.Monad.Freer.Writer
+
+import           Cardano.Prelude                 (NoUnexpectedThunks)
+import           Cardano.Slotting.Slot           (SlotNo (..), WithOrigin (..))
+import           Ouroboros.Network.Block         (HeaderHash, Point (..), StandardHash, pointSlot)
+import qualified Ouroboros.Network.Point         as Point
+
+import           Ledger                          (Block, Slot (..), Tx (..), TxId (..), txId)
+import qualified Ledger.Index                    as Index
+import           LedgerBytes                     (LedgerBytes (..))
+import           Plutus.SCB.EventLog
+import           Plutus.SCB.Events
+import           Plutus.SCB.Utils                (tshow)
+import           Wallet.Emulator.Chain           (ChainEffect (..), ChainState (..), chainNewestFirst, currentSlot,
+                                                  emptyChainState, handleChain, index, processBlock, queueTx)
+import qualified Wallet.Emulator.Chain           as EM
+import           Wallet.Emulator.NodeClient      (BlockValidated (..), NodeClientEffect (..), NodeClientEvent (..),
+                                                  NodeClientState (..), handleNodeClient)
+
+-- Block header
```

Comment:
Better haddocks for some of these. Also I think we could maybe do with a Note or something explaining how we use the node protocols.

Reply:
Generally all of this needs documentation!